omod/src/main/webapp/pages/results.gsp

```diff
@@ -32,21 +37,57 @@
                                           "bAutoWidth": false,
                                           "sDom": 'ft<"fg-toolbar ui-toolbar ui-corner-bl ui-corner-br ui-helper-clearfix datatables-info-and-pg "ip>'
                                       });
+
+        jq.fn.dataTableExt.ofnSearch['html'] = function (sData) {
+            return jq(sData).val();
+        };
+
+        jq('#rule_filter').change(function () {
+            resultTable.columns(0).search(jq(this).val()).draw();
+        });
+
+        jq("#patient_filter").autocomplete({
+            minLength: 0,
+            delay: 0,
+            source: Array.from(new Set(patients)),
+            select: function () {
+                resultTable.columns(3).search(jq(this).val()).draw();
+            }
+        });
+
     });
     <% } %>
 </script>
-<% if (rule == null) {
-%>
-<h2>${ui.message("dataintegrity.result.all.label")}</h2>
+
+<% if (rules == null) { %>
+    <h2>${ui.message("dataintegrity.result.all.label")}</h2>
+    <h3>${ui.message("dataintegrity.result.all.label")}</h3>
 <% } else { %>
-<h2>Results for ${rule.ruleName}</h2>
+    <h2>Results for ${rules.ruleName}</h2>
 <% } %>
+
+
+<% if (rules.size() != 0) { %>
+    <label for="rule_filter">Rule:</label>
+    <select id="rule_filter">
+        <option value="">Choose rule</option>
+        <% rules.each { %>
+        <option value="rule_${it.ruleId}">${ui.encodeHtmlContent(it.ruleName)}</option>
+        <% } %>
+    </select>
+
+    <label for="patient_filter">Choose Patient: </label>
+    <input id="patient_filter">
+<% } %>
+
+
 <% if (results.size() == 0) { %>
     There are no results to show. <a href="results.page?action=fireRules"><i class="icon-play"></i>&nbsp;Click here</a> to run all the validation rules, please note that it will take some time and slow down the server.
 <% } else { %>
 <table id="list-results" cellspacing="0" cellpadding="2" width="100%" class="dataTable">
     <thead>
     <tr>
+        <th style="display: none">Rule Id</th>
         <th>${ui.message("dataintegrity.rule.name")}</th>
         <th>${ui.message("dataintegrity.result.id")}</th>
```

Comment:
@ssmusoke , I guess all these attributes are defined in `messages.properties` and that file doesn't contain any attribute like `dataintegrity.result.entityid`. Please let me know in case I'm wrong. :)